packages/lesswrong/server/useractivities/cron.ts

```diff
@@ -0,0 +1,293 @@
+  // Go back to the start of the activity window, up to a maximum of ACTIVITY_WINDOW_HOURS in the past
+  const paddedStartDate = max([prevStartDate, new Date(updateEndDate.getTime() - (ACTIVITY_WINDOW_HOURS * 60 * 60 * 1000))]) ?? prevStartDate;
+  
+  // First case: Add rows for users who are newly active (zero padding the end as necessary)
+
+  // Prepare the new user data to be inserted in the UserActivities table
+  const newUsersData = cleanedActivityData
+    .filter(({ userOrClientId }) => !existingUserIds.includes(userOrClientId))
+    .map(({ userOrClientId, activityArray: activity_array }) => {
+      // paddedArray should be [...activity_array], plus zero padding going back to paddedStartDate:
+      const zeroPaddingLength = Math.floor((updateEndDate.getTime() - paddedStartDate.getTime()) / (60 * 60 * 1000)) - activity_array.length
+      const zeroPadding = Array(zeroPaddingLength).fill(0);
+      const paddedArray = [...activity_array, ...zeroPadding];
+      return { userOrClientId, paddedArray };
+    });
+
+  // Insert the new user data in a single query
+  if (newUsersData.length > 0) {
+    const newUsersDataChunked = chunk(1000, newUsersData)
+    console.log(`Inserting ${newUsersData.length} new rows into UserActivities table, in ${newUsersDataChunked.length} chunks`)
+    
+    for (const dataChunk of newUsersDataChunked) {
+      const placeholders = dataChunk.map((_, index) => `($${(index * 5) + 1}, $${(index * 5) + 2}, $${(index * 5) + 3}, $${(index * 5) + 4}, $${(index * 5) + 5}, '${visitorIdType}')`).join(', ');
+      const insertQuery = `
+        INSERT INTO "UserActivities" ("_id", "visitorId", "activityArray", "startDate", "endDate", "type")
+        VALUES ${placeholders}
+      `;
+      const queryParams = dataChunk.flatMap(({ userOrClientId, paddedArray }) => {
+        return [randomId(), userOrClientId, paddedArray, paddedStartDate.toISOString(), updateEndDate.toISOString()];
+      });
+      await dataDb.none(insertQuery, queryParams);
+    }
+  }
+
+  // Second case: Append the new activity to users' rows who were previously active.
+  // Note the truncation ([:${ACTIVITY_WINDOW_HOURS}]) to ensure that the array of activity is the correct length
+  const existingUsersData = cleanedActivityData.filter(({ userOrClientId }) => existingUserIds.includes(userOrClientId));
+  if (existingUsersData.length > 0) {
+    const existingUsersDataChunked = chunk(1000, existingUsersData)
+    console.log(`Updating ${existingUsersData.length} existing rows for which there is new activity, in ${existingUsersDataChunked.length} chunks`)
+    
+    for (const dataChunk of existingUsersDataChunked) {
+      const tempTableValues = dataChunk.map(({ userOrClientId, activityArray: activity_array }) => `('${userOrClientId}', ARRAY[${activity_array.join(', ')}])`).join(', ');
+      const updateQuery = `
+        WITH new_activity AS (
+          SELECT * FROM (VALUES ${tempTableValues}) AS t(user_id, activity_array)
```

Comment:
I feel like I'm not good enough at SQL for this. And you probably got this from GPT-4. Something something our new robot overlords.